### ISSUE TITLE
Config files must be of the same version of installed Varnish package

### ISSUE THREAD

**Issue by EmersonPrado** [claranet/puppet-varnish#40]
EDIT: this seems to only apply when `$varnish::addrepo = false`. Otherwise, this depends on whether the installer picks the package from Varnish repo or the distro ones.

Variable `$varnish::params::package_ensure` is set to `present`, causing Varnish to be installed in the latest version available. Variable `$varnish::params::varnish_version` is set to '3.0' or '4.0', depending on distro and version, so determining the version of the config files. This causes Varnish package and config files to have different versions in some situations. One example I hit in my tests is RHEL 6, which had Varnish installed with version 4 and configured with version 3 files.

As a workaround, I set `$varnish::varnish_version` to '4.0' in Hiera.

Suggestions:

- Class `varnish::config` could query the installed Varnish package version via a custom fact
- Variable `$varnish::varnish_version` could be set by checking the available installable version (like via `yum list`)

**Comment by EmersonPrado** [comment]
Provider `package` has methods called `latest` which, depending on child provider (apt, dpkg, etc.), can query either the installed or the latest available version for a given package.
I should learn how to write custom types.

**Comment by EmersonPrado** [comment]
Oops! Types don't return values. So it seems impossible to retrieve Varnish installed version from a custom type. Varnish installed version should be queried from a custom fact instead.
Problem: facts are collected **before** classes are executed, which means the installed version will be checked before Varnish is installed at all.

**Comment by EmersonPrado** [comment]
Added Work In Progress [PR 41](https://github.com/claranet/puppet-varnish/pull/41) with a proposed fix which takes into account the different combinations of relevant parameters I could think of.
Spec tests weren't updated, so they fail miserably.

**Comment by craigwatson** [comment]
Hi Emerson,

I'm not sure I understand the problem - if your specific system is using Varnish 4 via the OS default packages, and you need to override the default behaviour of the module, setting `varnish::varnish_version` in Hiera is the correct solution.

If the default package version within RHEL6 has changed from Varnish 3 to Varnish 4, we can adapt `params.pp` accordingly, but writing custom providers/facts shouldn't be necessary.

Please do correct me if any of the above is wrong.

Thanks,
Craig

**Comment by EmersonPrado** [comment]
Hi @craigwatson !

My problem is the package version and the config file version coming from different variables with different logic, opening the field for said mismatch. I think we need to make sure the package and config file match somehow. Maybe unifying the version parameter, still with default from `params`, then deriving both package and config file treatment from it.

Two important things to keep in mind:
1. RHEL usually connects to a local repository, like Satellite, what makes the OS default package version site-dependant. More specifically: dependant on site policy for repo update.
1. OS default package version is a moving target. True, doesn't change quite often. But does change, what puts a stress for mantainer to track each supported distro and for users to track module updates.

Hope this clears things up, but feel free to ask.

**Comment by craigwatson** [comment]
Two parameters are required for separate reasons:

- `varnish_version` is used to deploy config and external repos and is a more generic `major.minor` format
- `package_ensure` is passed directly to the `package` resource and allows pinning of a specific release version from the configured repository, be it OS or external

In theory we could extract the major/minor combination from `package_ensure` via regex to get `varnish_version` but this wouldn't address the edge case of needing the "latest" package from the 4.1 Packagecould repo, for example.

If your repositories are managed using Satellite and have different configurations per-site, I would strongly recommend using Hiera to specify the correct `varnish_version` and `package_ensure` settings for each site.

Specifying the OS-default package version is something I'm reconsidering, as CentOS requires EPEL in order to install Varnish, so the idea of a true "distro-default" doesn't really exist, and I'm unsure if installing EPEL is within scope of the module - although newer versions depend on `jemalloc` which is only available in EPEL anyway.

As Varnish 3 and 4.0 are [both EOL releases](https://varnish-cache.org/releases/), it would make logical sense to make an (opinionated) global default of 4.1 and throw away the idea of giving different defaults for each OS.

So, in summary - using Hiera is still the correct solution for this problem, but admittedly the module could make better use of both parameters.

**Comment by EmersonPrado** [comment]
> In theory we could extract the major/minor combination from `package_ensure` via regex to get `varnish_version`

You got the main idea: the issue is not the two variables itself, but the fact that they are determined independently. Something should tie them together to minimize the risk of mismatch.

> this wouldn't address the edge case of needing the "latest" package

Tough case indeed. I thought about the custom fact because it can query the "latest" actual version. But it has two issues: facts are collected early, before classes are evaluated at all, so any repo changes by Puppet can invalidate this data; and maybe it's overkill.
There is another edge case for `package_ensure => present`. In this case, the already installed version is kept, and the conf file should match it. A custom fact can check this fairly easily, but I really don't know if `package_ensure => present` is expected for this module.

> If your repositories are managed using Satellite and have different configurations per-site

In fact, I meant different Varnish versions for each RedHat customer which uses Satellite (or anything which manages local repos). Even with the same configurations, each company updates their Satellite repos at its own pace. So we can be sure that every time the latest Varnish version for RHEL changes, part of Satellite users will get a different version than others.

> I'm unsure if installing EPEL is within scope of the module

It really would conflict with anything that managed repos. But the module could call an external module to manage epel.

> using Hiera is still the correct solution for this problem, but admittedly the module could make better use of both parameters, and documentation on how to use them could be clearer

I would sum everything up with this suggestion:
1. Expose a single version parameter to the user, then get package and conf file versions from it. Or, as you said, expose `$package_ensure` parameter and and use it to form `$varnish_version`, which would not be a parameter anymore.
1. If `$package_ensure => 'present'`, form `$varnish_version` from the currently installed package. This would require a custom fact, though a fairly simple one.
1. If `$package_ensure => 'latest'`, either pin it at '4.1' or something meaningful for the OS, as you mentioned, or try to find the latest version available somehow - a custom fact is an option, but has the issue of early collection, as mentioned above.

Makes more sense?

**Comment by craigwatson** [comment]
If you are managing the repository outside of the module, it's logical to assume that you know which Varnish version you would be running and setting one or two parameters in Hiera should be fairly straightforward.

If repositories are managed outside of the module, it's the user's responsibility to set `varnish_version` correctly - although I accept that the parameters should be linked.

I'll close your PR, as a custom fact I think is overkill, but in the next few days I'll be adding some logic as part of a wider module refactor that will extract the `major.minor` version from the `package_ensure` parameter.
